src/components/chatbox/chatbox.tsx

```diff
@@ -0,0 +1,550 @@
+	// and we receive a 401 from the backend.
+	const { update, session } = useAuthentication()
+	const token = session?.accessToken
+
+	// The backend id of a conversation
+	const [conversationId, setConversationId] = useState('')
+	// The backend id of the most recently returned message
+	const [messageId, setMessageId] = useState('')
+
+	// Is the stream being read?
+	const [isReading, setIsReading] = useState(false)
+	// The streamed-in text
+	const [streamedText, setStreamedText] = useState('')
+	// Error text
+	const [errorText, setErrorText] = useState('')
+
+	type MutationParams = {
+		value: string
+		conversationId?: string
+		parentMessageId?: string
+	}
+
+	// Use useMutation to make a POST request more ergonomic
+	const mutation = useMutation<Response, Response, MutationParams>({
+		onMutate: async () => {
+			// clear previous response
+			setStreamedText('...')
+			setErrorText('')
+			setMessageId('')
+		},
+		onError: async (error) => {
+			console.log('onError', error)
+			switch (error.status) {
+				case 400:
+				case 401:
+				case 403: {
+					setErrorText(`${error.status} ${error.statusText}`)
+					break
+				}
+				case 429: {
+					const resetAtSec = mutation.error.headers.get('x-ratelimit-reset')
+					const resetAtMs = Number(resetAtSec) * 1000
+					const diffMs = resetAtMs - Date.now()
+
+					const errorMessage = `Too many requests. Please try again in ${ms(
+						diffMs,
+						{
+							long: true,
+						}
+					)}.`
+					setErrorText(errorMessage)
+					break
+				}
+				default: {
+					setErrorText(`${error.status} ${error.statusText}`)
+					break
+				}
+			}
+		},
+		mutationFn: async ({ value: task, conversationId, parentMessageId }) => {
+			// call our edge function which is capable of streaming
+			const request = async (jwt: string) => {
+				const res = await fetch('/api/chat/route', {
+					method: 'POST',
+					headers: {
+						'Content-Type': 'application/json',
+						Authorization: `Bearer ${token}`,
+					},
+					body: JSON.stringify({ task, conversationId, parentMessageId }),
+				})
+				return res
+			}
+
+			let response = await request(token)
+
+			if (response.ok) {
+				// messageId and conversationId are noops until we are ready for multi-message conversations
+				setConversationId(response.headers.get('x-conversation-id'))
+				setMessageId(response.headers.get('x-message-id'))
+				return response
+			} else {
+				// if the response is not ok
+				// if the response is a 401, we need to refresh the access token and retry once.
+				// otherwise, throw the response
+				if (response.status == 401) {
+					const { accessToken } = await update()
+					response = await request(accessToken)
+					if (response.ok) {
+						return response
+					}
+				}
+				throw response
+			}
+		},
+	})
+
+	// A stream reader
+	const [reader, setReader] =
+		useState<ReadableStreamDefaultReader<Uint8Array> | null>(null)
+	// Function to stop the stream
+	const stopStream = () => {
+		reader?.cancel()
```

Comment:
should this be a callback since it relies on the value of a `useState` hook?

Reply:
Do you mean, "should `stopStream` be memoized with `useCallback`" ?

(Assuming I understood your Q correctly...) I think it's technically fine as-is. Adding `useCallback` doesn't change functionality here.